parser.mly

```diff
@@ -27,23 +27,60 @@ let scope_annotation (mode, xs) =
 program: optional_newlines prog=program_code EOF { prog }
 
 program_code:
-| prog=list(instruction_line)
+| l1=instruction_line prog=list(instruction_line) fs=list(afunction)
   {
     let annotations, instructions = List.split prog in
-    [("main",
-      (Array.of_list instructions,
-       Array.of_list annotations))]
+    let a1, i1 = l1 in
+    { main = {
+          name = "main";
+          formals = [];
+          body = [("anon", Array.of_list (i1::instructions))];
+          annotations = [("anon", Array.of_list (a1::annotations))]; };
+      functions = fs;
+    }
+  }
+| f1=afunction fs=list(afunction)
+  {
+    let fs = f1 :: fs in
+    let main = (try List.find (fun {name = n} -> n = "main") fs with
+        | Not_found -> (Printf.printf ("missing main function\n"); exit 1)
```

Comment:
`exit 1` is a bit brutal (eg. it may stop the testsuite before tests are completed, although it might protect against it). From a design point of view, I think that having this check outside the parser (in a well-formedness verification pass of its own) would be better. For now, the simpler may be to just `raise Exit` -- it will fail badly (without a nicer-for-the-user error message), but it's not a panic. Or we could declare a new exception for that, but it should not be defined in the `Parser` or `Parse` modules.